test/LAMMPS_jll/.pkg/select_artifacts.jl

```diff
@@ -0,0 +1,22 @@
+# Note: This file is executed from `Pkg` in an isolated environment.
+# Warn: Since this is executed in a sandbox, the current preferences are not
+#       propagated and thus no decision can be made based on them.
+#       Use LazyArtifacts in this case.
+#       Fixed in https://github.com/JuliaLang/Pkg.jl/pull/2920
+
+push!(Base.LOAD_PATH, "@stdlib")
```

Comment:
This should be okay right? Otherwise we can't load `TOML` in all circumstances.

Reply:
Hmmm, in what circumstances does this not work?  I didn't need to do this for https://github.com/JuliaLang/Pkg.jl/blob/master/test/test_packages/AugmentedPlatform/.pkg/select_artifacts.jl

Reply:
I thought `TOML` was in `Base`? https://github.com/JuliaLang/julia/blob/b9432a8067a5bff7c1286c8e64b0572d5bd7a22c/base/toml_parser.jl#L3

Reply:
That's only a parser; if you want to write TOML output (like we do here) you need to load the stdlib.

Reply:
Funnily enough the testsuite here.

If I comment this out (on 1.7.1):

```
   Resolving package versions...
ERROR: LoadError: ArgumentError: Package TOML not found in current path:
- Run `import Pkg; Pkg.add("TOML")` to install the TOML package.

Stacktrace:
 [1] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:967
 [2] include(fname::String)
   @ Base.MainInclude ./client.jl:451
 [3] top-level scope
   @ none:5
in expression starting at /home/vchuravy/src/JLLWrappers/test/LAMMPS_jll/.pkg/select_artifacts.jl:8
  Stacktrace:
    [1] pipeline_error
      @ ./process.jl:531 [inlined]
    [2] read(cmd::Cmd)
      @ Base ./process.jl:418
    [3] collect_artifacts(pkg_root::String; platform::Base.BinaryPlatforms.Platform)
      @ Pkg.Operations /usr/share/julia/stdlib/v1.7/Pkg/src/Operations.jl:587
    [4] download_artifacts(env::Pkg.Types.EnvCache; platform::Base.BinaryPlatforms.Platform, julia_version::VersionNumber, verbose::Bool, io::Base.TTY)
      @ Pkg.Operations /usr/share/julia/stdlib/v1.7/Pkg/src/Operations.jl:614
    [5] develop(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}, new_git::Set{Base.UUID}; preserve::Pkg.Types.PreserveLevel, platform::Base.BinaryPlatforms.Platform)
```


Adding a `println(stderr, Base.LOAD_PATH)` shows that the LOAD_PATH at that time is:
```
["@", "/tmp/jl_NEdS9m"]
```

which is also the same at the time of the `Pkg.develop(PackageSpec(path=joinpath(@__DIR__, "LAMMPS_jll")))` call.

Reply:
Hmm, this probably isn't a problem for `Pkg` tests because `Pkg` itself loads `TOML`, so `@` expands to a project that has access to `TOML`.

Since literally all `select_artifact.jl` hooks will need to load `TOML` the stdlib, I think we should push onto `LOAD_PATH` as part of the generated code.

@KristofferC what do you think?

Reply:
I think the package should depend on TOML if it needs to load TOML. The whole point about not including `@stdlib` in the load path is to ensure dependencies are declared.

Reply:
Yeah, this is a weird gray area where we have a Pkg hook that is meant to spit out TOML, but it's executed in an environment that may not have `@stdlib` as a dependency.

It's not too onerous to require all these packages to declare a dependency on `TOML` though.  I think I could get behind that.

Reply:
Isn't problem here though , that this code is not executed in the environment of the package, so declaring a dependency won't do anything?

As we discussed the intention of the sandbox was to not require executing this code in the package environment, since at the point Pkg invokes this code, that environment might not be fully finalized. Which is why we can't depend on Preferences.

Reply:
So with https://github.com/JuliaLang/Pkg.jl/pull/2920/commits/9e770ea6fdade50aa885720e7085939b1a83f359 requiring `TOML` might work? Let me see if I can test that.

Reply:
No that is not sufficient.

https://github.com/JuliaLang/Pkg.jl/pull/2920/files#diff-6dfb4e40705b22a427451362ea83b4e3a763028e3dcaa3b9b1846aa1ec84370fR893

We set the `project` of the sandbox to the project that the user is currently working in. This is because we want those preferences to be picked up. The user might not need `TOML` in their current environment. 

An alternative to pushing `@stdlib` to the LOAD_PATH is to add `TOML` to the LAMMPS_jll dependencies and then
`push!(Base.LOAD_PATH, dirname(@__DIR__))`. But while working on https://github.com/JuliaLang/Pkg.jl/pull/2920 we run into issues were loading Preferences would fail in a fresh depot.

Reply:
> An alternative to pushing @stdlib to the LOAD_PATH is to add TOML to the LAMMPS_jll dependencies and then
push!(Base.LOAD_PATH, dirname(@__DIR__)).

Yep, that's what Kristoffer and I were talking about, and what I was referring to when I said `It's not too onerous to require all these packages to declare a dependency on TOML though. I think I could get behind that.`

> But while working on JuliaLang/Pkg.jl#2920 we run into issues were loading Preferences would fail in a fresh depot.

You shouldn't need to load `Preferences`; you should be able to just use `Base.get_preferences()`.

Reply:
Right, hypothetical world where we move TOML out of the stdlib we might encounter a similar problem.

Are we in agreement then, that we should add TOML as a dependency, and then `push!(Base.LOAD_PATH, dirname(@__DIR__))`?

Reply:
Why does https://github.com/JuliaLang/Pkg.jl/commit/9e770ea6fdade50aa885720e7085939b1a83f359 not suffice for loading TOML if it is listed as a dependency?

Reply:
Because we are setting `--project=$(Base.active_project())`. Which is not the same as `--project=LAMMPS_jll`.

We want to inherit preferences from the current user environment, but that does not need to have TOML declared.

Reply:
Right, right of course.  Okay yeah.  I'm happy with that `push!(Base.LOAD_PATH, dirname(@__DIR__))`.